### PR TITLE
First commit for Location Controller

### PR DESCRIPTION
**Pull request by flightblog**
- [ ] Support for iOS 7.

### CODE REVIEW

File: ExampleApplication/TWTSampleViewController.m [hunk 1]
```diff
@@ -52,6 +56,13 @@ - (void)viewDidLoad
     self.examples = [NSArray arrayWithContentsOfFile:toastSamplesURLString];
     
     [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:twt_cellIdentifier];
+
+    // TWTLocationController
+    // Update info.plist with either NSLocationWhenInUseUsageDescription and/or NSLocationAlwaysUsageDescription
+    [[TWTLocationController sharedInstance] fetchCurrentLocationWithCompletion:^(CLLocation *location, NSError *error) {
+        NSTimeInterval diff = [[NSDate date] timeIntervalSinceDate:location.timestamp];
```
Comment (prachigauriar) [inline ExampleApplication/TWTSampleViewController.m]: Not crazy about the name “diff.” What about something like, timeIntervalSinceUpdate or elapsedTime or whatever this is trying to tell us.

File: UIKit/Location Controller/TWTLocationController.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    TWTLocationControllerErrorPermissionDenied,
+    TWTLocationControllerErrorPermissionRestricted,
+    TWTLocationControllerErrorUpdateFailed,
+    TWTLocationControllerErrorUpdateTimedOut
+};
+
+extern NSString *const kTWTLocationControllerErrorDomain;
+
+
+typedef void(^TWTLocationControllerCompletion) (CLLocation *location, NSError *error);
```
Comment (prachigauriar) [inline UIKit/Location Controller/TWTLocationController.h]: Space after `void`. No space after `(^TWTLocationControllerCompletion *)`

File: UIKit/Location Controller/TWTLocationController.m [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+//  Created by Steve Foster on 2/3/15.
+//  Copyright (c) 2015 Two Toasters, LLC. All rights reserved.
+//
+
+#import "TWTLocationController.h"
+
+@import CoreLocation;
+
+static NSTimeInterval const kLocationTimeoutInterval = 60;
+NSString *const kLocationControllerErrorDomain = @"locationcontroller";
```
Comment (prachigauriar) [inline UIKit/Location Controller/TWTLocationController.m]: Error domains should be made public, else they’re not really useful. I’d call it `TWTLocationControllerErrorDomain` and make the string the same (“TWTLocationControllerErrorDomain”).

File: UIKit/Location Controller/TWTLocationController.m [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+
+@interface TWTLocationController () <CLLocationManagerDelegate>
+
+@property (nonatomic, strong) CLLocationManager *locationManager;
+@property (nonatomic, strong) CLLocation *location;
+@property (nonatomic, strong) NSMutableArray *requestBlocks;
```
Comment (prachigauriar) [inline UIKit/Location Controller/TWTLocationController.m]: Make these readonly. `requestBlocks` can be created once and emptied each time you use it.

File: UIKit/Location Controller/TWTLocationController.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+
+@interface TWTLocationController : NSObject
+
+@property (nonatomic, copy) TWTLocationControllerCompletion locationCompletion;
+
++ (instancetype)sharedInstance;
```
Comment (prachigauriar) [inline UIKit/Location Controller/TWTLocationController.h]: This probably should be named `defaultController`. `sharedInstance` implies that there’s only ever one and everyone should use it. `defaultController` implies that you can create as many as you want, but there’s a default one that you can use instead.

File: UIKit/Location Controller/TWTLocationController.h [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+extern NSString *const kTWTLocationControllerErrorDomain;
+
+
+typedef void(^TWTLocationControllerCompletion) (CLLocation *location, NSError *error);
+
+
+@interface TWTLocationController : NSObject
+
+@property (nonatomic, copy) TWTLocationControllerCompletion locationCompletion;
```
Comment (prachigauriar) [inline UIKit/Location Controller/TWTLocationController.h]: This shouldn’t be exposed, based on how I’m reading your code.

File: UIKit/Location Controller/TWTLocationController.m [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    }
+    return self;
+}
+
+
+#pragma mark - Fetch Location
+
+- (void)fetchCurrentLocationWithCompletion:(TWTLocationControllerCompletion)completion
+{
+    self.locationCompletion = completion;
```
Comment (prachigauriar) [inline UIKit/Location Controller/TWTLocationController.m]: This probably shouldn’t happen, right? You’ve got the requestBlocks variable that’s doing this instead.

File: UIKit/Location Controller/TWTLocationController.m [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+    self.requestBlocks = [[NSMutableArray alloc] init];
+    [self.requestBlocks addObject:completion];
```
Comment (prachigauriar) [inline UIKit/Location Controller/TWTLocationController.m]: By overwriting requestBlocks, aren’t you losing any previous requests. Isn’t the point of request blocks to track if there are any pending location updates?

File: UIKit/Location Controller/TWTLocationController.m (hunk 4 above)
Comment (prachigauriar) [inline UIKit/Location Controller/TWTLocationController.m]: I’d also rename requestBlocks to be pendingCompletionBlocks or something to that effect.

File: UIKit/Location Controller/TWTLocationController.m [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+#pragma mark - Location Request Timeout
+
+- (void)locationUpdateTimedOut:(NSTimer *)timer
+{
+    if ([self validateLocation:self.location accuracy:kCLLocationAccuracyThreeKilometers]) {
+        [self processCompletionBlocksWithLocation:self.location error:nil];
+        [self stopUpdatingLocation];
+    }
+    else {
+        NSLog(@"location timed out.");
```
Comment (prachigauriar) [inline UIKit/Location Controller/TWTLocationController.m]: Don’t log this. We’re a library, we pass up errors, not log them.

File: UIKit/Location Controller/TWTLocationController.m (hunk 8 above)
Comment (prachigauriar) [inline UIKit/Location Controller/TWTLocationController.m]: Basically, what you want to do is determine if you have any requests running. The obvious way to do this is to check if request blocks is empty or not. Unfortunately, you need to do that in a thread-safe manner, so you’ll need to use `@synchronized` or a GCD queue or something. We can talk about it in the office.

File: UIKit/Location Controller/TWTLocationController.m [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+- (void)locationManager:(CLLocationManager *)manager didFailWithError:(NSError *)error
+{
+    // TODO (foster): What to pass here as an error code.
+    [self failWithErrorCode:error.code underlyingError:error];
+}
+
+
+#pragma mark - Helpers
+
+- (void)updateLocationForAuthorizationStatus:(CLAuthorizationStatus)status
```
Comment (prachigauriar) [inline UIKit/Location Controller/TWTLocationController.m]: `updateLocationsWithAuthorizationStatus`.

File: UIKit/Location Controller/TWTLocationController.h [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    TWTLocationControllerAuthorizationStatusNotDetermined = 0,
+    TWTLocationControllerAuthorizationStatusRestricted,
+    TWTLocationControllerAuthorizationStatusDenied,
+    TWTLocationControllerAuthorizationStatusAuthorizedAlways,
+    TWTLocationControllerAuthorizationStatusAuthorizedWhenInUse,
+    TWTLocationControllerErrorPermissionRequired,
+    TWTLocationControllerErrorPermissionDenied,
+    TWTLocationControllerErrorPermissionRestricted,
+    TWTLocationControllerErrorUpdateFailed,
+    TWTLocationControllerErrorUpdateTimedOut
```
Comment (prachigauriar) [inline UIKit/Location Controller/TWTLocationController.h]: Some of these are error codes and some are authorization statuses. I don’t see why anyone would need an authorization status here that’s different than the CoreLocation one. As such, just make these all error codes. I’d also name this enum `TWTLocationControllerErrorCode` and prefix each error code with the same.

File: UIKit/Location Controller/TWTLocationController.m [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+{
+    switch ([CLLocationManager authorizationStatus]) {
+
+        case kCLAuthorizationStatusNotDetermined: {
+            [self failWithErrorCode:TWTLocationControllerAuthorizationStatusNotDetermined underlyingError:nil];
```
Comment (prachigauriar) [inline UIKit/Location Controller/TWTLocationController.m]: Should you fail in this case if you’re automatically going to request the appropriate authorization?

File: UIKit/Location Controller/TWTLocationController.m [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+            break;
+        }
+
+        case kCLAuthorizationStatusRestricted: {
+            [self failWithErrorCode:TWTLocationControllerAuthorizationStatusRestricted underlyingError:nil];
+            break;
+        }
+
+        case kCLAuthorizationStatusDenied: {
+            [self failWithErrorCode:TWTLocationControllerErrorPermissionDenied underlyingError:nil];
```
Comment (prachigauriar) [inline UIKit/Location Controller/TWTLocationController.m]: Seems like these cases should basically be wrapped up in a `TWTLocationControllerErrorCodeNotAuthorized` and you should pass in the actual status as a user info key (which should be documented and placed in the header).

File: UIKit/Location Controller/TWTLocationController.m [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+- (void)startUpdatingLocation
+{
+    [self.locationManager startUpdatingLocation];
+}
+
+
+- (void)stopUpdatingLocation
+{
+    [self.locationManager stopUpdatingLocation];
+}
```
Comment (prachigauriar) [inline UIKit/Location Controller/TWTLocationController.m]: Not entirely clear what the value of these methods are. Do you expect them to grow?

File: ExampleApplication/TWTSampleViewController.m [hunk 15]
```diff
@@ -52,6 +56,13 @@ - (void)viewDidLoad
     self.examples = [NSArray arrayWithContentsOfFile:toastSamplesURLString];
     
     [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:twt_cellIdentifier];
+
+    // TWTLocationController
+    // Update info.plist with either NSLocationWhenInUseUsageDescription and/or NSLocationAlwaysUsageDescription
+    [[TWTLocationController sharedInstance] fetchCurrentLocationWithCompletion:^(CLLocation *location, NSError *error) {
```
Comment (jnjosh) [inline ExampleApplication/TWTSampleViewController.m]: I suggest that this not be just something in the Sample View Controller. This really should be it's own sample view. See how the other samples are done in there. Hint, look into ToastExamples.plist

Comment (jnjosh) [inline ExampleApplication/TWTSampleViewController.m]: That means a very minimal UI too to show it working

File: UIKit/Location Controller/TWTLocationController.h (hunk 11 above)
Comment (jnjosh) [inline UIKit/Location Controller/TWTLocationController.h]: :+1:

File: UIKit/Location Controller/TWTLocationController.m [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+
+
+- (instancetype)init
+{
+    self = [super init];
+    if (self) {
+        self.locationManager = [[CLLocationManager alloc] init];
+        self.locationManager.delegate = self;
+        self.locationManager.desiredAccuracy = kCLLocationAccuracyHundredMeters;
```
Comment (jnjosh) [inline UIKit/Location Controller/TWTLocationController.m]: Is this the only way to configure the location controller? This is very limited. What if I want to use the location controller with other accuracies? This locks in how you have to use location manager.

File: UIKit/Location Controller/TWTLocationController.m [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+
+
+- (void)processCompletionBlocksWithLocation:(CLLocation *)location error:(NSError *)error
+{
+    NSArray *requestBlocks = [self.requestBlocks copy];
+    [self.requestBlocks removeAllObjects];
+
+    [[NSOperationQueue mainQueue] addOperationWithBlock:^{
+        for (TWTLocationControllerCompletion completion in requestBlocks) {
+            completion(self.location, error);
```
Comment (jnjosh) [inline UIKit/Location Controller/TWTLocationController.m]: Do you mean to be using self.location? Why does this method receive location?

File: ExampleApplication/TWTSampleViewController.m (hunk 15 above)
Comment (flightblog) [inline ExampleApplication/TWTSampleViewController.m]: Configured with `TWTLocationControllerSampleViewController.h/m`.

File: ExampleApplication/TWTSampleViewController.m (hunk 1 above)
Comment (flightblog) [inline ExampleApplication/TWTSampleViewController.m]: Used `timeIntervalSinceUpdate`.

File: UIKit/Location Controller/TWTLocationController.h (hunk 2 above)
Comment (flightblog) [inline UIKit/Location Controller/TWTLocationController.h]: Space removed.

File: UIKit/Location Controller/TWTLocationController.h (hunk 5 above)
Comment (flightblog) [inline UIKit/Location Controller/TWTLocationController.h]: Changed to `defaultController`.

File: UIKit/Location Controller/TWTLocationController.m (hunk 4 above)
Comment (flightblog) [inline UIKit/Location Controller/TWTLocationController.m]: Changed to `pendingCompletionBlocks'.

File: UIKit/Location Controller/TWTLocationController.m (hunk 16 above)
Comment (flightblog) [inline UIKit/Location Controller/TWTLocationController.m]: CLLocationManager is now exposed.

File: UIKit/Location Controller/TWTLocationController.m (hunk 9 above)
Comment (flightblog) [inline UIKit/Location Controller/TWTLocationController.m]: NSLog removed.

File: UIKit/Location Controller/TWTLocationController.m (hunk 10 above)
Comment (flightblog) [inline UIKit/Location Controller/TWTLocationController.m]: Updated.

File: UIKit/Location Controller/TWTLocationController.m (hunk 17 above)
Comment (flightblog) [inline UIKit/Location Controller/TWTLocationController.m]: Using the location that's being passed in.

File: UIKit/Location Controller/TWTLocationController.m (hunk 12 above)
Comment (flightblog) [inline UIKit/Location Controller/TWTLocationController.m]: Exposing `kCLAuthorizationStatusNotDetermined` makes it possible to present the user with a pre-premission alert and I also wanted to be complete in exposing all authorization statuses.

File: UIKit/Location Controller/TWTLocationController.m (hunk 16 above)
Comment (prachigauriar) [inline UIKit/Location Controller/TWTLocationController.m]: While that’s definitely the easiest solution, you need to be sure external users don’t become the delegate, as that will break this entire class.

You might be better served by having properties on your location controller which basically just pass through the settings to your CLLocationManager. I realize this is more code, but it seems like what you actually want.

**Comment by prachigauriar** [comment]
If you’ve made changes, check those in.

File: ExampleApplication/TWTSampleViewController.m (hunk 15 above)
Comment (flightblog) [inline ExampleApplication/TWTSampleViewController.m]: Sample View Controller created.

File: UIKit/Location Controller/TWTLocationController.h (hunk 6 above)
Comment (flightblog) [inline UIKit/Location Controller/TWTLocationController.h]: Fixed.

**Comment by Dfowj** [comment]
@flightblog This is moving to a standalone repository yes?

Once you've captured all the code/feedback above, lets close this PR. If you already have, go ahead and close this.

**Comment by flightblog** [comment]
Moved to a standalone project.

**Pull request closed without merging**